contracts/btc-light-client/src/bitcoin.rs

```diff
@@ -193,9 +215,60 @@ fn check_block_header_context(
         });
     }
 
+    let mtp = calculate_median_time_past(storage, header, prev_block_height + 1, pending_headers)?;
+    if header.time <= mtp {
+        return Err(HeaderError::TimeTooOld);
+    }
+
     Ok(())
 }
 
+/// Calculates the median time of the previous few blocks prior to the header (inclusive).
+// TODO: Added `header_height` to make the loop exit condition more robust when fetching previous
+// timestamps, since relying on `BlockHash::all_zeros()` isn't reliable with the current test data.
+// We may remove it in future if the test data is updated.
+fn calculate_median_time_past(
+    storage: &dyn Storage,
+    header: &BlockHeader,
+    header_height: u32,
+    pending_headers: &PendingHeaders,
+) -> Result<u32, HeaderError> {
+    let mut timestamps = Vec::with_capacity(MEDIAN_TIME_SPAN);
+
+    timestamps.push(header.time);
+
+    let mut block_hash = header.prev_blockhash;
+    let mut block_height = header_height - 1;
+
+    while timestamps.len() < MEDIAN_TIME_SPAN {
+        // Genesis block
+        if block_height == 0 {
+            break;
+        }
+
+        let header: BlockHeader = match get_header_by_hash(storage, block_hash.as_ref()) {
+            Ok(raw_header) => deserialize(raw_header.header.as_ref())?,
+            Err(_) => pending_headers
+                .get(&block_hash)
+                .map(|(header, _height)| header)
+                .cloned()
+                .ok_or(HeaderError::MissingHeader(block_hash))?,
```

Comment:
Does it mean if header is missing it will return error? do we have some check in contract that there is always at least 11 headers in store (for example after initialisation of contract) ?

Reply:
You're right, we should use as many headers as available, rather than erroring when fewer than 11 are present. I've rewritten the `calculate_median_time_past()` function accordingly. This change does require a conceptual shift in how we interact with storage.

Notably, the semantic of `get_header_by_hash()` has been updated to return `Result<Option<T>>` instead of `Result<T>`. This is because `None` can be a valid outcome in certain situations. Currently, many  of our storage APIs return an error when a value is missing, which might not always be semantically correct. Let me know if you have any thoughts on `get_header_by_hash()` and `expect_header_by_hash()`. cc @SebastianElvis

Reply:
>Notably, the semantic of get_header_by_hash() has been updated to return Result<Option<T>> instead of Result<T>. This is because None can be a valid outcome in certain situations. Currently, many of our storage APIs return an error when a value is missing, which might not always be semantically correct. Let me know if you have any thoughts on get_header_by_hash() and expect_header_by_hash()

👍 it makes sense to distinguish error and lack of element on type level